asf_public_discourse_home_decarbonisation/analysis/buildhub/extract_faqs.py

```diff
@@ -0,0 +1,245 @@
+        if "/" in sentence and sentence.count("/") > 2:
+            continue
+        # Find all matches in the sentence
+        potential_questions = re.findall(
+            question_pattern, sentence, flags=re.IGNORECASE
+        )
+        # Filter and add valid questions to the all_questions list
+        for question in potential_questions:
+            # Skip questions that are four words or less and contain a question word
+            words = question.split()
+            if len(words) <= num_of_words and any(
+                word.lower() in question_words for word in words
+            ):
+                continue
+            # Add sentence to all_questions if it's at least min_length long and not just question marks with optional spaces.
+            if len(question) >= min_length and not re.fullmatch(r"\?+\s*", question):
+                all_questions.append(question)
+
+    return all_questions
+
+
+def extract_idk(sentences: List[str]) -> Tuple[List[str], List[str]]:
+    """
+    Extracts and returns two lists based on a list of input sentences: one containing sentences with specific "I don't know" phrases, and another with these phrases removed.
+
+    The function searches for sentences that contain any of a predefined set of phrases indicating uncertainty or lack of knowledge (e.g., "I don’t know", "I do not know", "I do not know how to"). It then segregates these sentences into two lists: one preserving the original sentences, and another with the specific phrases removed.
+
+    This can be particularly useful for analysing text where it's important to both identify expressions of uncertainty and examine the context without these expressions.
+
+    Args:
+        sentences (List[str]): A list of sentences to be evaluated. Each item in the list should be a string representing a single sentence.
+
+    Returns:
+        Tuple[List[str], List[str]]:
+            - The first list contains sentences that include any of the predefined "I don't know" phrases.
+            - The second list contains the same sentences but with the "I don't know" phrases removed, allowing for analysis of the remaining sentence content.
+
+    Note:
+        - The function is case-insensitive when searching for the predefined phrases.
+        - The removal of phrases is done in a straightforward manner, which may not account for complex sentence structures or punctuation.
+    """
+    idk_phrases = []
+    sentences_without_inclusion = []
+    inclusion_phrases = [
+        "i don’t know ",
+        "i do not know ",
+        "i do not know how to ",
+    ]  # List of phrases to include
+    for sentence in sentences:
+        if any(phrase in sentence.lower() for phrase in inclusion_phrases):
+            idk_phrases.append(sentence)
+            for phrase in inclusion_phrases:
+                sentence = sentence.lower().replace(phrase, "")
+            sentences_without_inclusion.append(sentence)
```

Comment:
I am guessing the purpose of this was to extract the X in "... do not know how to X". However, if this is a bigger sentence, what we will get is ".... X" instead of just X.

Reply:
Also, as in the function above, we can have two functions: one where we extract the idk sentences and another where we extract the X :)

Reply:
I have split up the function and also it now just includes what follows on from the "I don't know" phrases.